### PR TITLE
(Wallet) Refactor fragment creation for onboarding adapter

### PR DESCRIPTION
**Pull request by simoarpe**
## Resolves https://github.com/brave/brave-browser/issues/37358

Perform refactoring on onboarding adapter moving logic of fragment creation into `createFragment(position)` method.
Fix terms of use checkbox state. 
Fix an issue affecting the backup screen that was throwing an illegal state exception.

## Submitter Checklist:

- [x] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### Terms of use checkbox issue

- Clear Wallet data, if any
- Navigate to Wallet section
- Tap on new Wallet button
- Select both checkboxes
- Tap the back arrow at top left side of the screen
-  Tap on restore Wallet button
- Observe term of use screen reset its state and checkboxes are not selected

#### Wallet Creation

- Clear Wallet data, if any
- Navigate to Wallet section
- Create a new Wallet
- Observe the Wallet creation flow works as expected

#### Wallet Restore

- Clear Wallet data, if any
- Navigate to Wallet section
- Restore a pre-existing Wallet
- Observe the Wallet restoration flow works as expected

#### Backup

- Open Wallet
- Tap on "Backup now" button
- Follow backup flow process
- Observe Wallet backup flow works as expected

### CODE REVIEW

File: android/java/org/chromium/chrome/browser/crypto_wallet/activities/BraveWalletActivity.java [hunk 1]
```diff
@@ -64,33 +47,14 @@ public class BraveWalletActivity extends BraveWalletBaseActivity implements OnNe
...
-        // Wallet.
-        PASSWORD_CREATION,
-        // Unlock action type triggered when accessing the locked Wallet.
-        UNLOCK,
-        // Restore action part of the onboarding flow, triggered when restoring a Wallet.
-        ONBOARDING_RESTORE,
-        // Restore action triggered outside onboarding flow on a pre-existing wallet.
-        RESTORE
-    }
-
```
Comment (simoarpe) [inline android/java/org/chromium/chrome/browser/crypto_wallet/activities/BraveWalletActivity.java:84]: Moved into `WalletOnboardingPagerAdapter` class.

File: android/java/org/chromium/chrome/browser/crypto_wallet/activities/BraveWalletActivity.java [hunk 2]
```diff
@@ -205,73 +186,6 @@ public ModalDialogManager getModalDialogManager() {
...
-            addBackupWalletSequence(navigationFragments, true);
-        }
-        mCryptoWalletOnboardingPagerAdapter.replaceWithNavigationItems(
-                navigationFragments, mCryptoWalletOnboardingViewPager.getCurrentItem() + 1);
-        mCryptoWalletOnboardingPagerAdapter.notifyDataSetChanged();
-
-        mCryptoWalletOnboardingViewPager.setCurrentItem(
-                mCryptoWalletOnboardingViewPager.getCurrentItem() + 1);
-    }
-
```
Comment (simoarpe) [inline android/java/org/chromium/chrome/browser/crypto_wallet/activities/BraveWalletActivity.java:274]: This entire logic has been replaced by a `position -> fragment` algorithm in `WalletOnboardingPagerAdapter#createFragment`. Doing so, the fragments are recycled when needed and their state is handled directly by the recycler, making them more memory efficient.

File: android/java/org/chromium/chrome/browser/crypto_wallet/fragments/UnlockWalletFragment.java [hunk 3]
```diff
@@ -93,32 +93,35 @@ public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceStat
...
+                }
             } else {
-                createBiometricPrompt();
-
+                mOnNextPage.enableBiometricPrompt();
+                showPasswordRelatedControls();
             }
-        } else if (mOnNextPage != null) {
-            mOnNextPage.enableBiometricPrompt();
-            showPasswordRelatedControls();
```
Comment (simoarpe) [inline android/java/org/chromium/chrome/browser/crypto_wallet/fragments/UnlockWalletFragment.java:121]: This logic was left untouched, I've just refactored `if` statements so the booleans are more readable.

File: android/java/org/chromium/chrome/browser/crypto_wallet/adapters/WalletOnboardingPagerAdapter.java [hunk 4]
```diff
@@ -0,0 +1,228 @@
+/* Copyright (c) 2021 The Brave Authors. All rights reserved.
```
Comment (SergeyZhukovsky) [inline android/java/org/chromium/chrome/browser/crypto_wallet/adapters/WalletOnboardingPagerAdapter.java:1]: 2024

File: android/java/org/chromium/chrome/browser/crypto_wallet/fragments/UnlockWalletFragment.java [hunk 5]
```diff
@@ -93,32 +93,35 @@ public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceStat
...
+        mBiometricUnlockWalletImage.setOnClickListener(
+                v -> {
+                    if (Utils.isBiometricAvailable(requireContext())) {
+                        // noinspection NewApi
+                        createBiometricPrompt();
+                    }
+                });
 
-        if (mOnNextPage != null && mOnNextPage.showBiometricPrompt()) {
-            if (Build.VERSION.SDK_INT < Build.VERSION_CODES.P
```
Comment (SergeyZhukovsky) [inline android/java/org/chromium/chrome/browser/crypto_wallet/fragments/UnlockWalletFragment.java:111]: I think biometric is not available on API <=28. We should probably leave that check and show only a password way in that case.

File: android/java/org/chromium/chrome/browser/crypto_wallet/fragments/UnlockWalletFragment.java [hunk 6]
```diff
@@ -201,8 +204,7 @@ private void showPasswordRelatedControls() {
         mUnlockButton.setVisibility(View.VISIBLE);
         mUnlockWalletRestoreButton.setVisibility(View.VISIBLE);
         mUnlockWalletTitle.setVisibility(View.VISIBLE);
-        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P
```
Comment (SergeyZhukovsky) [inline android/java/org/chromium/chrome/browser/crypto_wallet/fragments/UnlockWalletFragment.java:204]: the same about biometric and API <=28.

File: android/java/org/chromium/chrome/browser/crypto_wallet/adapters/WalletOnboardingPagerAdapter.java (hunk 4 above)
Comment (simoarpe) [reply android/java/org/chromium/chrome/browser/crypto_wallet/adapters/WalletOnboardingPagerAdapter.java:1]: I don't know why it's not indicated by GitHub UI (it usually does it), but this file has been renamed from `CryptoWalletOnboardingPagerAdapter` I haven't implemented it from scratch, shall I update the copyright year anyway?

File: android/java/org/chromium/chrome/browser/crypto_wallet/fragments/UnlockWalletFragment.java (hunk 5 above)
Comment (simoarpe) [reply android/java/org/chromium/chrome/browser/crypto_wallet/fragments/UnlockWalletFragment.java:111]: Correct, biometric is only available on devices running Android API 29 and upwards.
I've removed that statement as we were doing this check twice: `Utils.isBiometricAvailable(Context)` in the line below contains already a check that returns false if `VERSION.SDK < VERSION_CODES.P`.

https://github.com/brave/brave-core/blob/927cabbefb76849be5dd032adc813c14f74c7380/android/java/org/chromium/chrome/browser/crypto_wallet/util/Utils.java#L1028-L1032

File: android/java/org/chromium/chrome/browser/crypto_wallet/fragments/UnlockWalletFragment.java (hunk 6 above)
Comment (simoarpe) [reply android/java/org/chromium/chrome/browser/crypto_wallet/fragments/UnlockWalletFragment.java:204]: https://github.com/brave/brave-core/pull/23127/files#r1574378595

File: android/java/org/chromium/chrome/browser/crypto_wallet/fragments/UnlockWalletFragment.java (hunk 5 above)
Comment (SergeyZhukovsky) [reply android/java/org/chromium/chrome/browser/crypto_wallet/fragments/UnlockWalletFragment.java:111]: ok cool, I missed that!

**Review by SergeyZhukovsky — APPROVED** [review]
++

**Pull request merged**